docs(development): guide to create/edit presets (#13165)
renovatebot/renovate: docs/development/creating-editing-renovate-presets.md

# Creating/editing Renovate presets

Renovate comes with default presets that you can find in the `lib/config/presets/internal` directory.
You can suggest changes to the presets with a pull request.

Follow the rules below to increase the chance that your pull request gets merged.

## General rules

> 1. Only create/edit presets for problems which can not be fixed upstream

Author: I think we should nudge our users/contributors into fixing the problem upstream, that way bad stuff gets fixed for all downstream users not just Renovate bot itself.

It also means less maintenance work for us. 😄

1. The internal preset must be helpful for most Renovate users

### Specific rules

#### Group presets

> File location: `lib/config/presets/internal/group.ts`

Author: Do we need to mention the precise path to the preset?
This can get outdated quick if we refactor our code later on, so maybe it's better to leave it out?

Rules:

> 1. Only group dependencies that _must_ be updated together

Reviewer: Actually we are not this strict.

We have the `group:monorepo` preset, although that extends from the `monorepo` one so the rules for that can be stated there.

The `group:recommended` preset is for related dependencies which aren't from a monorepo but which usually need to be updated together as separate PRs may each break.

Finally, any other `group:*` presets can be added here if they are beneficial to a wide number of users. They don't need to be added to `group:recommended`, meaning that users will "opt in" to them one by one and not get them automatically from `config:base`, which includes `group:monorepo` and `group:recommended`.

Author: I've copy/pasted your response into the document now. 😉

#### Replacement presets

File location: `lib/config/presets/internal/replacements.ts`

Rules:

> 1. Only suggest a replacement PR when the user is on the latest `major` release of the deprecated dependency, use the `matchCurrent` config option to control when the replacement PR is allowed
> 1. No big version jumps, update to the first stable version of the replacement dependency

Reviewer: ```suggestion
1. Replacement PRs should ideally propose a replacement only once the user is on a compatible version, by specifying a compatible `matchCurrentVersion` constraint.
1. If no compatible replacement upgrade is possible, it's acceptable to propose an incompatible one (e.g. a major version upgrade).
1. Replacements should update the user to the first recommended version of the new dependency and not include any new changes - whether breaking or not - if they can be avoided.
```

#### Monorepo presets

File location: `lib/config/presets/internal/monorepo.ts`

Rules:

> 1. First check if the problem can be fixed in the source monorepo itself, sometimes the monorepo is not configured properly

Reviewer: ```suggestion
1. The primary use case of monorepo presets is identifying packages from the same origin source repository which should be updated together.
1. Packages from the same repository which are developed and versioned independently do not need to be grouped as a monorepo, but in many cases we still do.
1. Packages from separate repositories but which are released together and dependent on each other may also be added to the "monorepo" definitions even if not strictly true.
```

> 1. First check if the problem can be fixed in the source monorepo itself, sometimes the monorepo is not configured properly

Author: I don't know what the common causes are for monorepo misconfiguration.
I think Renovate relies on some fields in the `package.json` file to understand that it's a monorepo?

Should we explain how to fix monorepo config problems upstream, or is that out of scope for Renovate documentation?

Reviewer: Monorepo presets aren't usually about fixing a problem

# Creating/editing Renovate presets

Renovate comes with default presets that you can find in the `lib/config/presets/internal` directory.
You can suggest changes to the presets with a pull request.

Follow the rules below to increase the chance that your pull request gets merged.

## General rules

> 1. Only create/edit presets for problems which can not be fixed upstream

Reviewer: ```suggestion
1. Avoid creating presets for problems which can be fixed upstream
```

1. The internal preset should be helpful for a significant number of Renovate users

### Specific rules

#### Group presets

We have multiple kinds of `group:` presets, with different rules.

##### Rules for `group:monorepo` preset

1. Only group dependencies that _must_ be updated together

##### Rules for `group:recommended` presets

1. The `group:recommended` preset is for related dependencies which aren't from a monorepo but which usually need to be updated together as separate PRs may each break

##### Rules for `group:*` presets

1. Finally, any other `group:*` presets can be added if they are beneficial to a wide number of users

> 2. They don't need to be added to `group:recommended`, meaning that users will "opt in" to them one by one and not get them automatically from `config:base`, which includes `group:monorepo` and `group:recommended`

Author: ```suggestion
1. They don't need to be added to `group:recommended`, meaning that users will "opt in" to them one by one and not get them automatically from `config:base`, which includes `group:monorepo` and `group:recommended`
```

You're right! Fixing it now. 😄

#### Replacement presets

Rules:

> 1. Replacement PRs should ideally propose a replacement only once the user is on a compatible version, by specifying a compatible `matchCurrentVersion` constraint

Author: Maybe we should mention to limit the range if there are multiple `replacements` / package renames? We needed to "narrow the version range" for the `joi` to `@hapi/joi` replacement, so that people didn't get downgraded to a version of `joi` they upgraded past already...

See PR https://github.com/renovatebot/renovate/pull/13205

Reviewer: I think that one's probably a rarely one where there were multiple major versions each with both packages?

Author: That whole `joi`/`@hapi/joi` and then back to `joi` thing seems to be an edgecase.

Maybe we can leave out the `, by specifiying a compatible matchCurrentVersion` step in the instructions?

1. If no compatible replacement upgrade is possible, it's acceptable to propose an incompatible one (e.g. a major version upgrade)

> 1. Replacements should update the user to the first recommended version of the new dependency and not include any new changes - whether breaking or not - if they can be avoided

Reviewer 2: Wouldn't rule 1 and 3 cause a three-stage approach?:

1. Update to the latest version of the initial package
2. Switch to the first version of the replacement package
3. Update to the latest version of the replacement package

So for example: `ant:ant:1.4.1` => `ant:ant:1.7.0` => `org.apache:ant:1.7.0` => `org.apache.ant:ant:1.10.12`

Would that be the preferable outcome of replacements?

Reviewer: Yes, in general it would cause that. But we can be flexible if someone thinks there's a better way. What I really want to avoid is *breaking* replacement changes when a smoother alternative is possible

Author: That is what we want!

We want users to:

- First be within the latest `major` range of the old package
- Then they'll get an `replacement` PR from Renovate to update them to the earliest possible version of the replacement package
- After that is merged, they'll get Renovate PRs for updates as normal.

Author: Are you happy with the current text then? I'm not sure if I need to edit the text or not based on this discussion.
